README.md
# thesis_introduction

Branch |[![Travis CI logo](pics/TravisCI.png)](https://travis-ci.org)                                                                                         
-------|------------------------------------------------------------------------------------------------------------------------------------------------------------
master |[![Build Status](https://travis-ci.org/richelbilderbeek/thesis_introduction.svg?branch=master)](https://travis-ci.org/richelbilderbeek/thesis_introduction) 
develop|[![Build Status](https://travis-ci.org/richelbilderbeek/thesis_introduction.svg?branch=develop)](https://travis-ci.org/richelbilderbeek/thesis_introduction)

My thesis' Introduction.



NEWS.md
# News

Newest versions at top.

## thesis_introduction 0.3 (2020-03-16)

### NEW FEATURES

  * Travis CI build fails upon spelling error
  
### MINOR IMPROVEMENTS

  * Implemented feedback from @rsetienne

### BUG FIXES

  * None

### DEPRECATED AND DEFUNCT

  * None



create_figures.R
# Creates the figures for the introduction
setwd("~/GitHubs/thesis_introduction")

suppressMessages(library(ggtree))


#
#      +--------x   A
#   ---+
#      |     +------- B
#      +-----+
#            +------- C
#
#  -+--+-----+--+---- (time)
#   t0 t1    t2 t3
#
ggtree::ggtree(
  ape::read.tree(text = "((C:2,B:2):1, A:2):1;"),
  ladderize = FALSE
) + ggtree::geom_tiplab(cex = 10) +
  ggtree::geom_rootedge() +
  ggtree::geom_treescale(width = 6, x = -2, fontsize = 0) +
  ggplot2::geom_vline(xintercept = -1, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 0, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 1, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 2, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 3, lty = "dotted") +
  ggplot2::geom_label(
    data = data.frame(x = seq(-1, 3), y = 0, label = paste0("t", seq(0,4))),
    aes(x = x, y = y, label = label),
    label.size = 0.0,
    nudge_x = 0,
    nudge_y = -0.185,
    cex = 10
  ) +
  ggplot2::geom_label(
    data = data.frame(x = 4, y = 0, label = "t"),
    aes(x = x, y = y, label = label),
    label.size = 0.0,
    nudge_x = 0.1,
    nudge_y = 0.0,
    cex = 10
  ) + ggplot2::ggsave(filename = "t_3.png", width = 7, height = 4)



#
#     ----- FUCA
#
#  ---+---- (time)
#     t0
#

ggtree::ggtree(
  ape::read.tree(text = "(FUCA:0):1;"),
  ladderize = FALSE
) + ggtree::geom_tiplab(cex = 10) +
  ggtree::geom_rootedge(rootedge = 1) +
  ggtree::geom_treescale(width = 6, x = -2, fontsize = 0) +
  ggplot2::geom_vline(xintercept = -1, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 0, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 1, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 2, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 3, lty = "dotted") +
  ggplot2::geom_label(
    data = data.frame(x = seq(-1, 3), y = 0, label = paste0("t", seq(0,4))),
    aes(x = x, y = y, label = label),
    label.size = 0.0,
    nudge_x = 0,
    nudge_y = -0.185,
    cex = 10
  ) +
  ggplot2::geom_label(
    data = data.frame(x = 4, y = 0, label = "t"),
    aes(x = x, y = y, label = label),
    label.size = 0.0,
    nudge_x = 0.1,
    nudge_y = 0.0,
    cex = 10
  ) + ggplot2::ggsave(filename = "t_0.png", width = 7, height = 4)


#
#      +---- A
#   ---+
#      +---- B
#
#  -+--+--- (time)
#   t0 t1
#

ggtree::ggtree(
  ape::read.tree(text = "(B:1,A:1);"),
  ladderize = FALSE
) + ggtree::geom_tiplab(cex = 10) +
  ggtree::geom_rootedge(1) +
  ggtree::geom_treescale(width = 6, x = -2, fontsize = 0) +
  ggplot2::geom_vline(xintercept = -1, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 0, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 1, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 2, lty = "dotted") +
  ggplot2::geom_vline(xintercept = 3, lty = "dotted") +
  ggplot2::geom_label(
    data = data.frame(x = seq(-1, 3), y = 0, label = paste0("t", seq(0,4))),
    aes(x = x, y = y, label = label),
    label.size = 0.0,
    nudge_x = 0,
    nudge_y = -0.185,
    cex = 10
  ) +
  ggplot2::geom_label(
    data = data.frame(x = 4, y = 0, label = "t"),
    aes(x = x, y = y, label = label),
    label.size = 0.0,
    nudge_x = 0.1,
    nudge_y = 0.0,
    cex = 10
  ) + ggplot2::ggsave(filename = "t_1.png", width = 7, height = 4)


scripts/create_nltts.R
# Creates the nLTT figures

library(pirouette)
library(ggplot2)

root_folder <- path.expand("~/GitHubs/thesis/introduction")
set.seed(314)

true_tree <- ape::read.tree(text = "(((A:8, B:8):1, C:9):1, ((D:8, E:8):1, F:9):1);")
twin_tree <- pirouette::create_twin_tree(
  phylogeny = true_tree,
  twinning_params = create_twinning_params(rng_seed = 314)
)

################################################################################
# Plotting
################################################################################
grDevices::png(filename = file.path(root_folder, "true_tree.png"), width = 800, height = 600)
# ape::plot.phylo(phylogeny, cex = 2.0, edge.width = 2.0)
ggtree::ggtree(true_tree, size = 2, col = "red") + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()

grDevices::png(filename = file.path(root_folder, "twin_tree.png"), width = 800, height = 600)
# ape::plot.phylo(phylogeny, cex = 2.0, edge.width = 2.0)
ggtree::ggtree(twin_tree, size = 2, col = "blue") + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()

grDevices::png(filename = file.path(root_folder, "nltt.png"), width = 800, height = 600)
par(mar = c(5.1, 4.1, 4.1, 2.1) + 5)
nLTT::nltts_plot(c(true_tree), col = "red", lwd = 7, main = "nLTTs", cex.lab = 2.5, cex.main = 4, cex.axis = 2.5)
nLTT::nltts_plot(c(twin_tree), replot = TRUE, col = "blue", lwd = 7)
grDevices::dev.off()



scripts/create_yule_and_bd_tree_figures.R
# Creates the figures for the Yule and BD trees

library(pirouette)
library(ggplot2)
library(ggtree)

set.seed(314)
tree <- pirouette::create_yule_tree(n_taxa = 10)

grDevices::png(filename = "~/GitHubs/thesis_introduction/yule_tree.png", width = 800, height = 600)
par(mar = c(5.1, 4.1, 4.1, 2.1) + 5)
ggtree::ggtree(tree, size = 2, col = "black")  + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()

grDevices::png(filename = "~/GitHubs/thesis_introduction/yule_tree_nltt.png", width = 800, height = 600)
par(mar = c(5.1, 4.1, 4.1, 2.1) + 5)
nLTT::nltts_plot(c(tree), col = "black", lwd = 7, main = "nLTT", cex.lab = 2.5, cex.main = 4, cex.axis = 2.5)
grDevices::dev.off()


set.seed(315)
tree <- pirouette::create_bd_tree(n_taxa = 10)

grDevices::png(filename = "~/GitHubs/thesis_introduction/bd_tree.png", width = 800, height = 600)
par(mar = c(5.1, 4.1, 4.1, 2.1) + 5)
ggtree::ggtree(tree, size = 2, col = "black")  + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()

grDevices::png(filename = "~/GitHubs/thesis_introduction/bd_tree_nltt.png", width = 800, height = 600)
par(mar = c(5.1, 4.1, 4.1, 2.1) + 5)
nLTT::nltts_plot(c(tree), col = "black", lwd = 7, main = "nLTT", cex.lab = 2.5, cex.main = 4, cex.axis = 2.5)
grDevices::dev.off()



scripts/effect_of_assumptions_on_humans.R
# Creates the posteriors for humans

library(pirouette)
library(ggplot2)

root_folder <- path.expand("~/GitHubs/thesis_introduction")
setwd(root_folder)
set.seed(314)

# Read primates
nexus_file <- beastier::get_beast2_example_filename("Primates.nex")
alignment <- ape::as.DNAbin(ape::read.nexus.data(nexus_file))

# Create a FASTA filename
fasta_filename <- tempfile()
ape::write.FASTA(x = alignment, file = fasta_filename)
readLines(fasta_filename)

# Inference models
mcmc <- create_mcmc(chain_length = 1e6, store_every = 1e3)
n_states <- 1 + (mcmc$chain_length / mcmc$store_every)
halfway_index <- (mcmc$chain_length / mcmc$store_every) / 2


# Run babette with Yule model
inference_model_yule <- create_inference_model(
  tree_prior = create_yule_tree_prior(),
  mcmc = mcmc
)
out_yule <- bbt_run_from_model(
  fasta_filename = fasta_filename,
  inference_model = inference_model_yule
)

png("primates_yule.png")
babette::plot_densitree(
  out_yule[[1]][halfway_index:n_states],
  scaleX = TRUE,
  main = "Yule",
  type = "phylogram",
  width = 2,
  alpha = 10 / n_states
)
dev.off()

# Birth-Death
inference_model_bd <- create_inference_model(
  tree_prior = create_bd_tree_prior(),
  mcmc = mcmc
)
out_bd <- bbt_run_from_model(
  fasta_filename = fasta_filename,
  inference_model = inference_model_bd
)
png("primates_bd.png")
babette::plot_densitree(
  out_bd[[1]][halfway_index:n_states],
  scaleX = TRUE,
  main = "BD",
  type = "phylogram",
  width = 2,
  alpha = 10 / n_states
)
dev.off()

# Run babette with CEP
inference_model_cep <- create_inference_model(
  tree_prior = create_cep_tree_prior(),
  mcmc = mcmc
)
out_cep <- bbt_run_from_model(
  fasta_filename = fasta_filename,
  inference_model = inference_model_cep
)
png("primates_cep.png")
babette::plot_densitree(
  out_cep[[1]][halfway_index:n_states],
  scaleX = TRUE,
  main = "CEP",
  type = "phylogram",
  width = 2,
  alpha = 10 / n_states
)
dev.off()


if (1 == 2) {

  install.packages("PhyloOrchard", repos="http://R-Forge.R-project.org")
  require(PhyloOrchard)
  data(BinindaEmondsEtAl2007)
  BinindaEmondsEtAl2007 <- .compressTipLabel(BinindaEmondsEtAl2007)
  phangorn::densiTree(BinindaEmondsEtAl2007, type = "phylogram", col="red", main = "Hello")



  png("~/title_around.png")
  plot(1:10, main = "my_title")
  dev.off()

  png("~/title_missing.png")
  phangorn::densiTree(c(ape::rcoal(n = 10)), main = "my_title")
  dev.off()


  phangorn::densiTree(c(ape::rcoal(n = 10)), main = "Hello")
  ?plot
  ?phangorn::densiTree
}


scripts/effect_of_assumptions_on_bd_trees.R
# Effect of species tree prior on inference

library(pirouette)
library(ggplot2)

set.seed(314)
root_folder <- path.expand("~/GitHubs/thesis_introduction")
setwd(root_folder)

true_trees <- TreeSim::sim.bd.taxa(n = 9, numbsim = 1, lambda = 10.0, mu = 9.0, complete = FALSE)
true_tree <- true_trees[[1]]
true_tree$tip.label <- paste0("t", seq(1, 10))
ape::plot.phylo(true_tree)

# Inference models
mcmc <- create_mcmc(chain_length = 1e6, store_every = 1e3)
n_states <- 1 + (mcmc$chain_length / mcmc$store_every)
halfway_index <- (mcmc$chain_length / mcmc$store_every) / 2

alignment_params <- create_test_alignment_params(
  root_sequence = create_blocked_dna(1000)
)

mrca_prior <- create_mrca_prior(
  mrca_distr = create_normal_distr(mean = 1.0, sigma = 0.0001),
  is_monophyletic = TRUE
)

pir_params_yule <- create_test_pir_params(
  alignment_params = alignment_params,
  experiments = list(
      create_test_gen_experiment(
      inference_model = create_inference_model(
        tree_prior = create_yule_tree_prior(),
        mcmc = mcmc,
        mrca_prior = mrca_prior
      )
    )
  )
)
pir_params_bd <- create_test_pir_params(
  alignment_params = alignment_params,
  experiments = list(
      create_test_gen_experiment(
      inference_model = create_inference_model(
        tree_prior = create_bd_tree_prior(),
        mcmc = mcmc,
        mrca_prior = mrca_prior
      )
    )
  )
)


out_yule <- pirouette::pir_run(
  phylogeny = true_tree,
  pir_params = pir_params_yule
)

out_bd <- pirouette::pir_run(
  phylogeny = true_tree,
  pir_params = pir_params_bd
)

png("bd_tree_yule_prior.png")
babette::plot_densitree(
  tracerer::parse_beast_trees(
    pir_params_yule$experiments[[1]]$beast2_options$output_trees_filenames
  )[halfway_index:n_states],
  alpha = 10 / n_states,
  type = "phylogram",
  consensus = sort(true_tree$tip.label)
)
dev.off()

png("bd_tree_bd_prior.png")
babette::plot_densitree(
  tracerer::parse_beast_trees(
    pir_params_bd$experiments[[1]]$beast2_options$output_trees_filenames
  )[halfway_index:n_states],
  alpha = 10 / n_states,
  type = "phylogram",
  consensus = sort(true_tree$tip.label)
)
dev.off()


scripts/create_workflow_pics_long_alignment.R
# Creates the figures for the workflow image in the documentation

library(pirouette)
library(ggplot2)

dna_sequence_length <- 40
root_folder <- path.expand(
  paste0(
    "~/GitHubs/thesis_introduction/scripts/",
    dna_sequence_length
  )
)
set.seed(314)

phylogeny  <- ape::read.tree(text = "(((A:8, B:8):1, C:9):1, ((D:8, E:8):1, F:9):1);")
n_taxa <- ape::Ntip(phylogeny)
crown_age <- max(ape::branching.times(phylogeny))
################################################################################
# Use png here, to make Peregrine fail fast
################################################################################
grDevices::png(filename = file.path(root_folder, "phylogeny.png"), width = 800, height = 600)
# ape::plot.phylo(phylogeny, cex = 2.0, edge.width = 2.0)
ggtree::ggtree(phylogeny, size = 2) + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()


pir_params <- create_pir_params(
  alignment_params = create_alignment_params(
    root_sequence = pirouette::create_blocked_dna(length = dna_sequence_length),
    mutation_rate = 0.5 / crown_age
  ),
  twinning_params = create_twinning_params()
)

################################################################################
# Settings to run on Peregrine cluster
################################################################################
if (1 == 2) {
  pir_params$alignment_params$fasta_filename <- file.path(root_folder, "alignment.fasta")
  pir_params$experiments[[1]]$beast2_options$input_filename <- file.path(root_folder, "beast2_input.xml")
  pir_params$experiments[[1]]$beast2_options$output_log_filename <- file.path(root_folder, "beast2_output.log")
  pir_params$experiments[[1]]$beast2_options$output_trees_filenames <- file.path(root_folder, "beast2_output.trees")
  pir_params$experiments[[1]]$beast2_options$output_state_filename <- file.path(root_folder, "beast2_output.xml.state")
  pir_params$experiments[[1]]$errors_filename <- file.path(root_folder, "error.csv")
}
################################################################################

errors <- pir_run(phylogeny = phylogeny, pir_params = pir_params)

################################################################################
# trees
################################################################################

grDevices::png(filename = file.path(root_folder, "phylogeny.png"), width = 800, height = 600)
# ape::plot.phylo(phylogeny, cex = 2.0, edge.width = 2.0)
ggtree::ggtree(phylogeny, size = 2, col = "red") + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()


grDevices::png(filename = file.path(root_folder, "phylogeny_twin.png"), width = 800, height = 600)
# ape::plot.phylo(ape::read.tree(pir_params$twinning_params$twin_tree_filename), cex = 2.0, edge.width = 2.0)
ggtree::ggtree(ape::read.tree(pir_params$twinning_params$twin_tree_filename), size = 2, col = "orange") + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()


################################################################################
# alignment
################################################################################

n_mutations <- count_n_mutations(
  ape::read.FASTA(pir_params$alignment_params$fasta_filename),
  root_sequence = pir_params$alignment_params$root_sequence
)
print(paste("n_mutations", n_mutations))
print(paste("n_mutations per taxon", n_mutations / n_taxa))

grDevices::png(filename = file.path(root_folder, "alignment.png"), width = 800, height = 300)
ape::image.DNAbin(
  ape::read.FASTA(file = pir_params$alignment_params$fasta_filename),
  grid = TRUE,
  show.bases = FALSE,
  legend = FALSE,
  cex.lab = 2.0,
  cex.axis = 2.0
)
grDevices::dev.off()

# Add annotation
fasta <- readLines(pir_params$alignment_params$fasta_filename)
fasta <- c(">", pir_params$alignment_params$root_sequence, fasta)
fasta_filename <- file.path(root_folder, "alignment_with_root.fas")
writeLines(text = fasta, con = fasta_filename)
png_filename <- file.path(root_folder, "alignment_with_root.png")
grDevices::png(filename = png_filename, width = 800, height = 300)
ape::image.DNAbin(
  ape::read.FASTA(file = fasta_filename),
  grid = TRUE,
  show.bases = FALSE,
  legend = FALSE,
  cex.lab = 2.0,
  cex.axis = 2.0
)
grDevices::dev.off()

grDevices::png(filename = file.path(root_folder, "alignment_twin.png"), width = 800, height = 300)
ape::image.DNAbin(
  ape::read.FASTA(file = pir_params$twinning_params$twin_alignment_filename),
  grid = TRUE,
  show.bases = FALSE,
  legend = FALSE,
  cex.lab = 2.0,
  cex.axis = 2.0
)
grDevices::dev.off()

################################################################################
# posteriors
################################################################################

grDevices::png(filename = file.path(root_folder, "densitree.png"), width = 1000, height = 800)
babette::plot_densitree(
  phylos = tracerer::parse_beast_trees(pir_params$experiments[[1]]$beast2_options$output_trees_filenames),
  alpha = 0.01,
  consensus = LETTERS[6:1],
  cex = 6.0,
  scaleX = TRUE,
  scale.bar = FALSE
)
grDevices::dev.off()

grDevices::png(filename = file.path(root_folder, "densitree_twin.png"), width = 1000, height = 800)
babette::plot_densitree(
  phylos = tracerer::parse_beast_trees(to_twin_filename(pir_params$experiments[[1]]$beast2_options$output_trees_filenames)),
  alpha = 0.01,
  consensus = LETTERS[6:1],
  cex = 6.0,
  scaleX = TRUE,
  scale.bar = FALSE
)
grDevices::dev.off()

################################################################################
# histogram of errors
################################################################################
df_errors <- data.frame(error = utils::read.csv(pir_params$experiments[[1]]$errors_filename)$x)
df_errors_twin <- data.frame(error = utils::read.csv(to_twin_filename(pir_params$experiments[[1]]$errors_filename))$x)

ggplot2::ggplot(
  df_errors,
  aes(x = error)
) + geom_histogram(binwidth = 0.01) +
  ggplot2::theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 30, face = "bold")
  ) + ggsave(file.path(root_folder, "errors.png"))

ggplot2::ggplot(
  df_errors_twin,
  aes(x = error)
) + geom_histogram(binwidth = 0.01) +
  ggplot2::theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 30, face = "bold")
  ) + ggsave(file.path(root_folder, "errors_twin.png"))

ggplot2::ggplot(
  df_errors,
  aes(x = "", y = error)
) + geom_violin() +
  xlab("") +
  scale_y_continuous(breaks = seq(0.0, 1.0, by = 0.02)) +
  ggsave(file.path(root_folder, "errors_violin.png"))

ggplot2::ggplot(
  df_errors_twin,
  aes(x = "", y = error)
) + geom_violin() +
  xlab("") +
  scale_y_continuous(breaks = seq(0.0, 1.0, by = 0.02)) +
  ggsave(file.path(root_folder, "errors_violin_twin.png"))


################################################################################
# Create ML tree from alignment
################################################################################
alignment <- ape::read.FASTA(pir_params$alignment_params$fasta_filename)

dm  <- phangorn::dist.ml(alignment)
tree_upgma  <- phangorn::upgma(dm)
tree_nj  <- phangorn::NJ(dm)
tree_fastme  <- ape::fastme.bal(dm)
tree_bionj  <- ape::bionj(dm)

grDevices::png(filename = file.path(root_folder, "phylogeny_upgma.png"), width = 800, height = 600)
ggtree::ggtree(tree_upgma, size = 2, col = "blue") + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()

grDevices::png(filename = file.path(root_folder, "phylogeny_nj.png"), width = 800, height = 600)
ggtree::ggtree(tree_nj, size = 2) + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()

grDevices::png(filename = file.path(root_folder, "phylogeny_fastme.png"), width = 800, height = 600)
ggtree::ggtree(tree_fastme, size = 2) + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()

grDevices::png(filename = file.path(root_folder, "phylogeny_bionj.png"), width = 800, height = 600)
ggtree::ggtree(tree_bionj, size = 2) + ggtree::geom_tiplab(size = 16) + ggplot2::theme(plot.margin = unit(c(1,1,1,1), "cm"))
grDevices::dev.off()





################################################################################
# nLTTs
################################################################################
grDevices::png(filename = file.path(root_folder, "nltt_true_and_upgma.png"), width = 800, height = 600)
par(mar = c(5.1, 4.1, 4.1, 2.1) + 5)
nLTT::nltts_plot(c(phylogeny), col = "red", lwd = 7, main = "nLTTs", cex.lab = 2.5, cex.main = 4, cex.axis = 2.5)
nLTT::nltts_plot(c(tree_upgma), replot = TRUE, col = "blue", lwd = 7)
grDevices::dev.off()

